load_testing/common_flows.py

```diff
@@ -0,0 +1,154 @@
+from locust_helpers import authenticity_token, resp_to_dom, random_cred
+import foney
+import os
+
+# Common flows that are used across tests
+
+"""
+*** Sign In Flow ***
+"""
+def do_sign_in(context):
+
+    # This should match the number of users that were created for the DB with the rake task
+    num_users = os.environ.get('NUM_USERS')
+    credentials = random_cred(num_users)
+
+    with context.client.get('/', catch_response=True) as resp:
+        # If you're already logged in, it'll redirect to /account.
+        # We need to handle this or you'll get all sorts of
+        # downstream failures.
+        if '/account' in resp.url:
+            print("You're already logged in. We're going to quit sign-in.")
+            return resp
+
+        dom = resp_to_dom(resp)
+        token = authenticity_token(dom)
+
+        if not token:
+            resp.failure(
+                "Not a sign-in page. Current URL is {}.".format(resp.url)
+            )
+
+    resp = context.client.post(
+        resp.url,
+        data={
+            'user[email]': credentials['email'],
+            'user[password]': credentials['password'],
+            'authenticity_token': authenticity_token(dom),
+            'commit': 'Submit',
+        },
+        catch_response=True
+    )
+    dom = resp_to_dom(resp)
+    code = dom.find("#code").attr('value')
+
+    if not code:
+        resp.failure(
+            """
+            No 2FA code found.
+            Make sure {} is created and can log into the IDP
+            """.format(credentials)
+        )
+        return
+
+    code_form = dom.find("form[action='/login/two_factor/sms']")
+    
+    resp = context.client.post(
+        code_form.attr('action'),
+        data={
+            'code': code,
+            'authenticity_token': authenticity_token(dom),
+            'commit': 'Submit'
+        }
+    )
+
+    # We're not checking for post-login state here,
+    # as it will vary depending on the SP.
+    resp.raise_for_status()
+
+    return resp
+
+
+"""
+*** Sign Up Flow ***
+"""
+def do_sign_up(context):
+  fake = Faker()
+  phone_numbers = foney.phone_numbers()
+  default_password = "salty pickles"
+
+  # GET the new email page
+  resp = context.client.get("/sign_up/enter_email")
+  dom = resp_to_dom(resp)
+
+  # Post fake email and get confirmation link (link shows up in "load test mode")
+  new_email = "test+{}@test.com".format(fake.md5())
+  resp = context.client.post("/sign_up/enter_email",
+            data={
+              'user[email]': new_email,
+              'authenticity_token': authenticity_token(dom),
+            },
+            catch_response=True)
+
+  with resp as confirm_resp:
+      dom = resp_to_dom(confirm_resp)
+      try:
+          confirmation_link = dom.find("#confirm-now")[0].attrib['href']
+      except Exception as err:
+          resp.failure(
+              "Could not find CONFIRM NOW link, is IDP enable_load_testing_mode: 'true' ?" + err)
+          return 
```

Comment:
what's the point of this `return`? Is it the same result if it's not there?

Reply:
I'm so bad at python is the reason :) Thank you for all of these great pieces of feedback, I'll give them all a go.